// @genkit-ai/google-genai/lib/googleai/deep-research.mjs
import { modelActionMetadata, z } from "genkit";
import {
  modelRef
} from "genkit/model";
import { backgroundModel as pluginBackgroundModel } from "genkit/plugin";
import {
  ensureToolIds,
  fromInteraction,
  toInteractionTool,
  toInteractionTurn
} from "../common/interaction-converters.js";
import { cleanSchema } from "../common/utils.js";
import {
  cancelInteraction,
  createInteraction,
  getInteraction
} from "./client.js";
import {
  calculateApiKey,
  checkApiKey,
  checkModelName,
  extractVersion,
  modelName
} from "./utils.js";
const DeepResearchConfigSchema = z.object({
  apiKey: z.string().describe("Override the API key provided at plugin initialization.").optional(),
  baseUrl: z.string().describe(
    "Overrides the plugin-configured or default baseUrl, if specified."
  ).optional(),
  apiVersion: z.string().describe(
    "Overrides the plugin-configured or default apiVersion, if specified."
  ).optional(),
  thinkingSummaries: z.enum(["AUTO", "NONE"]).describe("Whether to include thought summaries in the response.").optional(),
  previousInteractionId: z.string().describe("The ID of the previous interaction, if any.").optional(),
  store: z.boolean().describe(
    "Whether to store the response and request for later retrieval."
  ).optional(),
  responseModalities: z.array(z.enum(["TEXT", "IMAGE", "AUDIO"])).describe("The modalities to be used in response.").optional()
}).passthrough();
function commonRef(name, info, configSchema = DeepResearchConfigSchema) {
  return modelRef({
    name: `googleai/${name}`,
    configSchema,
    info: info ?? {
      supports: {
        multiturn: true,
        media: false,
        tools: false,
        toolChoice: false,
        systemRole: false,
        output: ["text"],
        longRunning: true
      }
    }
  });
}
const GENERIC_MODEL = commonRef("deep-research");
const KNOWN_MODELS = {
  "deep-research-pro-preview-12-2025": commonRef(
    "deep-research-pro-preview-12-2025"
  )
};
function isDeepResearchModelName(value) {
  return !!value?.startsWith("deep-research-");
}
function model(version, config = {}) {
  const name = checkModelName(version);
  if (KNOWN_MODELS[name]) {
    return KNOWN_MODELS[name].withConfig(config);
  }
  return modelRef({
    name: `googleai/${name}`,
    config,
    configSchema: DeepResearchConfigSchema,
    info: { ...GENERIC_MODEL.info }
  });
}
function listActions(models) {
  return models.filter((m) => isDeepResearchModelName(modelName(m.name))).filter((m) => !m.description || !m.description.includes("deprecated")).map((m) => {
    const ref = model(m.name);
    return modelActionMetadata({
      name: ref.name,
      info: ref.info,
      configSchema: ref.configSchema
    });
  });
}
function listKnownModels(options) {
  return Object.keys(KNOWN_MODELS).map(
    (name) => defineModel(name, options)
  );
}
function defineModel(name, pluginOptions) {
  checkApiKey(pluginOptions?.apiKey);
  const ref = model(name);
  const clientOptions = {
    apiVersion: pluginOptions?.apiVersion,
    baseUrl: pluginOptions?.baseUrl,
    customHeaders: pluginOptions?.customHeaders
  };
  return pluginBackgroundModel({
    name: ref.name,
    ...ref.info,
    configSchema: ref.configSchema,
    async start(request) {
      const {
        apiKey: apiKeyConfig,
        baseUrl,
        apiVersion,
        thinkingSummaries,
        previousInteractionId,
        store,
        responseModalities,
        ...rest
      } = request.config || {};
      const apiKey = calculateApiKey(pluginOptions?.apiKey, apiKeyConfig);
      const newClientOptions = {
        ...clientOptions,
        apiKey,
        baseUrl: baseUrl || clientOptions.baseUrl,
        apiVersion: apiVersion || clientOptions.apiVersion
      };
      const messages = structuredClone(request.messages);
      for (const message of messages) {
        if (message.role === "system") {
          message.role = "user";
        }
      }
      const tools = [];
      if (request.tools?.length) {
        for (const tool of request.tools) {
          tools.push(toInteractionTool(tool));
        }
      }
      let responseMimeType = request.output?.contentType;
      let responseFormat = void 0;
      if (request.output?.format === "json" || request.output?.contentType === "application/json") {
        responseMimeType = "application/json";
        if (request.output.schema) {
          responseFormat = cleanSchema(request.output.schema);
        }
      }
      let responseModalitiesConverted;
      if (responseModalities) {
        responseModalitiesConverted = responseModalities.map(
          (m) => m.toLowerCase()
        );
      }
      const req = {
        agent: extractVersion(ref),
        input: ensureToolIds(messages).map(toInteractionTurn),
        background: true,
        agent_config: {
          type: "deep-research",
          thinking_summaries: thinkingSummaries?.toLowerCase()
        },
        previous_interaction_id: previousInteractionId,
        store,
        tools: tools.length ? tools : void 0,
        response_mime_type: responseMimeType,
        response_format: responseFormat,
        response_modalities: responseModalitiesConverted,
        ...rest
      };
      const response = await createInteraction(apiKey, req, newClientOptions);
      return fromInteraction(response, newClientOptions);
    },
    async check(operation) {
      const storedOptions = operation.metadata?.clientOptions;
      const apiKey = storedOptions?.apiKey || calculateApiKey(pluginOptions?.apiKey, void 0);
      const checkOptions = {
        ...clientOptions,
        ...storedOptions
      };
      const response = await getInteraction(apiKey, operation.id, checkOptions);
      return fromInteraction(response, checkOptions);
    },
    async cancel(operation) {
      const storedOptions = operation.metadata?.clientOptions;
      const apiKey = storedOptions?.apiKey || calculateApiKey(pluginOptions?.apiKey, void 0);
      const cancelOptions = {
        ...clientOptions,
        ...storedOptions
      };
      const response = await cancelInteraction(
        apiKey,
        operation.id,
        cancelOptions
      );
      return fromInteraction(response, cancelOptions);
    }
  });
}
const TEST_ONLY = { KNOWN_MODELS };
export {
  DeepResearchConfigSchema,
  TEST_ONLY,
  defineModel,
  isDeepResearchModelName,
  listActions,
  listKnownModels,
  model
};
